network/src/dms2.rs

```diff
@@ -0,0 +1,670 @@
+}
+
+struct DmsWrapper<S: Storage> {
+    #[allow(clippy::type_complexity)]
+    /// This is an `Option` because we have to explicitly drop the server
+    /// (it could live forever in the RPC server (`axum`) otherwise)
+    dms: Arc<parking_lot::RwLock<Option<Arc<RwLock<DistributedMessageSet<S>>>>>>,
+}
+
+#[async_trait]
+impl<S: Storage> DistributedMessageSetRpcInterface for DmsWrapper<S> {
+    async fn get_message(
+        &self,
+        dms_key: DmsKey,
+        knowns: Vec<Hash256>,
+    ) -> Result<Vec<RawMessage>, String> {
+        let dms = Arc::clone(
+            self.dms
+                .read()
+                .as_ref()
+                .ok_or_else(|| "server terminated".to_owned())?,
+        );
+        let mut messages = dms
+            .read()
+            .await
+            .read_messages()
+            .await
+            .map_err(|e| e.to_string())?;
+        let dms_key_ = dms.read().await.key.clone();
+        if dms_key != dms.read().await.key {
+            return Err(format!("key mismatch: requested {dms_key}, but {dms_key_}"));
+        }
+        let knowns: HashSet<_> = knowns.into_iter().collect();
+        let messages: Vec<_> = messages
+            .drain(..)
+            .filter(|m| !knowns.contains(&m.to_hash256()))
+            .map(RawMessage::from_message)
+            .collect();
+        Ok(messages)
+    }
+
+    async fn add_messages(&self, dms_key: DmsKey, messages: Vec<RawMessage>) -> Result<(), String> {
+        let dms = Arc::clone(
+            self.dms
+                .read()
+                .as_ref()
+                .ok_or_else(|| "server terminated".to_owned())?,
+        );
+        let dms_key_ = dms.read().await.key.clone();
+        if dms_key != dms.read().await.key {
+            return Err(format!("key mismatch: requested {dms_key}, but {dms_key_}"));
+        }
+        for message in messages {
+            let message = message.into_message().map_err(|e| e.to_string())?;
+            dms.write()
+                .await
+                .add_message(message)
+                .await
+                .map_err(|e| e.to_string())?;
+        }
+        Ok(())
+    }
+}
+
+struct DummyFilter;
+
+impl MessageFilter for DummyFilter {
+    fn filter(&self, _message: &Message) -> Result<(), String> {
+        Ok(())
+    }
+}
+
+/// Runs a DMS server. This function will block the current thread.
+pub async fn serve<S: Storage>(
+    dms: Arc<RwLock<DistributedMessageSet<S>>>,
+    network_config: ServerNetworkConfig,
+) -> Result<(), Error> {
+    let port_key = format!("dms-{}", dms.read().await.key);
+    let port = network_config
+        .ports
+        .get(&port_key)
+        .ok_or_else(|| eyre!(format!("`ports` has no field of {port_key}")))?;
+
+    let dms_ = Arc::clone(&dms);
+    let rpc_task = async move {
+        let wrapped_dms = Arc::new(parking_lot::RwLock::new(Some(dms_)));
```

Comment:
Why do you clone `dms` here? I mean, we can just consume it.

Reply:
Is it for other fetching methods like gossipping I guess?

Reply:
Yes, as I copied the code from the old DMS. I will fix it.